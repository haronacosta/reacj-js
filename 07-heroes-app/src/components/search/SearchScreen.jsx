import React, { useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { getHerobyName } from '../../selectors/getHerobyName';
import HeroCard from '../hero/HeroCard';

const SearchScreen = () => {
  const navigate = useNavigate();

  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const initialSearch = {
    searchValue: q,
  };

  const [{ searchValue }, handleInputChange] = useForm(initialSearch);

  const heroesFiltered = useMemo(() => getHerobyName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();

    navigate(`?q=${searchValue}`);
  };

  return (
    <>
      <h4>Search a hero</h4>

      <hr />

      <div className='row'>
        <div className='col-5'>
          <h5>Search</h5>
          <hr />

          <form onSubmit={handleSearch}>
            <input
              className='form-control mb-4'
              type='text'
              name='searchValue'
              target='searchValue'
              placeholder='Search hero'
              autoComplete='off'
              value={searchValue}
              onChange={handleInputChange}
            />

            <button type='submit' className='btn btn-primary'>
              Search...
            </button>
          </form>
        </div>

        <div className='col-7'>
          <h5>Results</h5>

          <hr />

          {q === '' ? (
            <div className='alert alert-info'>
              <p>Please, type a name to search a hero.</p>
            </div>
          ) : (
            heroesFiltered.length === 0 && (
              <div className='alert alert-danger'>Not results</div>
            )
          )}

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};

export default SearchScreen;
