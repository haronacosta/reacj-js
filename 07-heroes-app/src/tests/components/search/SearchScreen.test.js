import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas en el <SearchScreen />', () => {
  test('Debe regresar el componente correctamente', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info p').text().trim()).toBe(
      'Please, type a name to search a hero.'
    );
  });

  test('Debe mostrar a batman y el input con el queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('input').prop('value')).toBe('batman');
  });

  test('Debe de mostrar el mensaje de no hay resultados cuando se pase batman123', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('input').prop('value')).toBe('batman123');
    expect(wrapper.find('.alert-danger').text().trim()).toBe('Not results');
  });

  test('Debe llamar el navigate a la nueva pantalla', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchValue',
        value: 'batman',
      },
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {},
    });

    expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
  });
});
