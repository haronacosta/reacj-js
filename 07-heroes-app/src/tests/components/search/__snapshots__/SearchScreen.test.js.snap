// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe de mostrar el mensaje de no hay resultados cuando se pase batman123 1`] = `
Array [
  <h4>
    Search a hero
  </h4>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-4"
          name="searchValue"
          onChange={[Function]}
          placeholder="Search hero"
          target="searchValue"
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h5>
        Results
      </h5>
      <hr />
      <div
        className="alert alert-danger"
      >
        Not results
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el <SearchScreen /> Debe mostrar a batman y el input con el queryString 1`] = `
Array [
  <h4>
    Search a hero
  </h4>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-4"
          name="searchValue"
          onChange={[Function]}
          placeholder="Search hero"
          target="searchValue"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h5>
        Results
      </h5>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row g-4"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-title"
              >
                Batman
              </div>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                Detective Comics #27
              </p>
              <a
                className="pl-4 btn btn-outline-success"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s info
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el <SearchScreen /> Debe regresar el componente correctamente 1`] = `
Array [
  <h4>
    Search a hero
  </h4>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-4"
          name="searchValue"
          onChange={[Function]}
          placeholder="Search hero"
          target="searchValue"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h5>
        Results
      </h5>
      <hr />
      <div
        className="alert alert-info"
      >
        <p>
          Please, type a name to search a hero.
        </p>
      </div>
    </div>
  </div>,
]
`;
